src/gui/updater/appimageupdater.cpp

```diff
@@ -0,0 +1,202 @@
+                _logStatusMessages();
+                return false;
+            }
+
+            _logStatusMessages();
+            return updateAvailable;
+        } catch (const std::exception &e) {
+            _logStatusMessages();
+            qCCritical(lcUpdater) << "Checking for update failed:" << e.what();
+            return false;
+        }
+    }
+
+    void startUpdateInBackground() noexcept
+    {
+        // monitor progress and log status messages
+        auto *timer = new QTimer(this);
+
+        timer->setInterval(100ms);
+
+        connect(timer, &QTimer::timeout, this, [=]() {
+            _logStatusMessages();
+
+            if (_updater.isDone()) {
+                emit finished(!_updater.hasError());
+                timer->stop();
+            }
+        });
+
+        _updater.start();
+        timer->start();
+    }
+
+signals:
+    void finished(bool successfully);
+
+private:
+    appimage::update::Updater _updater;
+};
+
+} // namespace
+
+AppImageUpdater::AppImageUpdater(const QUrl &url)
+    : OCUpdater(url)
+{
+}
+
+
+bool AppImageUpdater::handleStartup()
+{
+    // nothing to do, update will be performed while app is running, if anything
+    return false;
+}
+
+void AppImageUpdater::versionInfoArrived(const UpdateInfo &info)
+{
+    const auto &currentVersion = Version::versionWithBuildNumber();
+    const auto newVersion = QVersionNumber::fromString(info.version());
+
+    if (info.version().isEmpty() || currentVersion >= newVersion) {
+        qCInfo(lcUpdater) << "Client is on latest version!";
+        setDownloadState(UpToDate);
+        return;
+    }
+
+    auto settings = ConfigFile::makeQSettings();
+
+    const auto seenVersionString = settings.value(seenVersionC).toString();
+    if (!seenVersionString.isEmpty()) {
+        if (QVersionNumber::fromString(seenVersionString) >= QVersionNumber::fromString(info.version())) {
+            qCInfo(lcUpdater) << "Update" << seenVersionString << "was skipped previously by user";
+            setDownloadState(UpToDate);
+            return;
+        }
+    }
+
+    const auto appImageUpdaterShim = AppImageUpdaterShim::makeInstance(info.downloadUrl(), this);
+
+    if (appImageUpdaterShim == nullptr) {
+        setDownloadState(DownloadFailed);
+        return;
+    }
+
+    if (!appImageUpdaterShim->isUpdateAvailable()) {
+        qCCritical(lcUpdater) << "Update server reported that update is available, but AppImageUpdate disagrees, aborting";
+        setDownloadState(DownloadFailed);
+        return;
+    }
+
+    auto dialog = new Ui::AppImageUpdateAvailableDialog(currentVersion, newVersion, ocApp()->gui()->settingsDialog());
+
+    connect(dialog, &Ui::AppImageUpdateAvailableDialog::skipUpdateButtonClicked, this, [newVersion]() {
+        qCInfo(lcUpdater) << "Update" << newVersion << "skipped by user";
+
+        auto settings = ConfigFile::makeQSettings();
+        settings.setValue(seenVersionC, newVersion.toString());
```

Comment:
This is all very similar to the nsis code, can it be refactored?

Reply:
The only similarities I see are fetching and storing the seen version as well as checking whether the client is up to date or there is no version info available. I moved the former into two static utility functions within `OCUpdater`, but I think moving half of the up-to-date check into some helper doesn't solve any actual problem.